### PR TITLE
USHIFT-432: CSI image / manifest rebase script

### PR DESCRIPTION
**Pull request by copejon**
**Which issue(s) this PR addresses**:

[USHIFT-432](https://issues.redhat.com//browse/USHIFT-432)

### CODE REVIEW

File: pkg/release/release.go
Comment (dhellmann) [inline pkg/release/release.go:28]: I think most of the whitespace changes in this file are triggered by this comment. Maybe Jira is a better place to track the todo?

File: scripts/auto-rebase/rebase_odf-lvm.sh
Comment (dhellmann) [inline scripts/auto-rebase/rebase_odf-lvm.sh:73]: We go to a lot of trouble to parse and update the release.go file here and in the main rebase script. What if we wrote a YAML file to the assets directory instead, and had release.go load and parse it? We would want to make the same change for all images, so we should not hold up this PR over the discussion.

Comment (copejon) [reply scripts/auto-rebase/rebase_odf-lvm.sh:73]: Yeah, I think you're right.  We go through a lot of trouble to do hacky string replacements in the release*.go files.  We'd probably be better off reusing our bindata pipeline to embed the release image data in the binary from a yaml source, rather than all the shell-driven string substitutions we do now.

Comment (fzdarsky) [reply scripts/auto-rebase/rebase_odf-lvm.sh:79]: In general, though, I agree this script should align more with the other rebase script in style and reuse code.

**Comment by copejon** [comment]
Failing due to missing namespace fields in operator bundles' manifests.  Will fix.

File: assets/components/lvms/topolvm-node-metrics_v1_service.yaml
Comment (ggiguash) [inline assets/components/lvms/topolvm-node-metrics_v1_service.yaml:17]: Is this something we had missed in manual rebase that was found by the automation?

File: assets/release/release-aarch64.json
Comment (ggiguash) [inline assets/release/release-aarch64.json:23]: We currently have the same images without "topolvm-" prefix. There seems to be some duplication here. Should the old names be removed?

File: assets/release/release-x86_64.json
Comment (ggiguash) [inline assets/release/release-x86_64.json:23]: Same as above. There seems to be some duplication for csi- and topolvm-csi- images.

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:52]: ARM images are available as part of the same manifest. 
See https://catalog.redhat.com/software/containers/lvms4/topolvm-rhel8/63972de3adcb55263891b983?architecture=arm64&container-tabs=gti

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:73]: Can we also pass the output file as an argument, so that it's not defined in two places?

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:96]: The file will "always" be created. Should we check that the file is not empty?

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:109]: Should we log a warning message here? Is this a "normal" situation?

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:122]: Let's remove this?

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:123]: This can be done by `$(ls -1 .)` without the need for `sed`

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:134]: Can you explain what we're trying to do here? We are taking a manifest and overwriting it eventually?

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:164]: Should we remove this?

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:1080]: This may have been a debugging option, but would we ever want to update just csi images? We do not have special option for ovn, etc., so why csi is special?

**Review by ggiguash — CHANGES_REQUESTED** [review]

File: assets/components/lvms/topolvm-node-metrics_v1_service.yaml
Comment (ggiguash) [reply assets/components/lvms/topolvm-node-metrics_v1_service.yaml:17]: Is it necessary to backport this fix to 4.12 release?

File: assets/release/release-aarch64.json
Comment (copejon) [reply assets/release/release-aarch64.json:23]: The topolvm name here reflects the container names in the upstream OCP manifests and means we don't have to maintain a mapping of OCP -> uShift images.

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:52]: Ah, yes.  This is stale, will remove.

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:73]: +1

File: assets/components/lvms/topolvm-node-metrics_v1_service.yaml
Comment (copejon) [reply assets/components/lvms/topolvm-node-metrics_v1_service.yaml:17]: no, the previous version was also pulled from the operator bundle, so this change will have come from upstream topolvm

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:109]: This check exists to filter out topolvm images used in OCP (like snapshotter) that are not used in microshift.  It's using the current release-*.json images list as the source of truth for this.

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:96]: Good catch. I'll adjust it to check for content rather than file presence.

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:1080]: We're free to remove this.  It was initially for debugging, but b/c LVMS isn't tethered to OCP z-streams (i.e. isn't tracked by the ocp-release image), I figured it'd be useful to have should we need to pull in OCP y-stream compatible versions ad hoc.

File: assets/release/release-aarch64.json
Comment (ggiguash) [reply assets/release/release-aarch64.json:23]: So, we should probably remove all the csi-* images from the JSON files?

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [reply scripts/auto-rebase/rebase.sh:1080]: I think I'd rather remove this option to keep it similar with the other components.

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:134]: The manifests in the operator bundle don't specify namespace, and MicroShift doesn't inherently assume a namespace for LVMS workloads.  This command feeds the downloaded manifests through `oc` to inject the namespace and outputs new yamls.  Using `oc` is nice b/c it won't specify a namespace on a non-namespaced object.

File: assets/release/release-aarch64.json
Comment (copejon) [reply assets/release/release-aarch64.json:23]: My mistake, I hadn't caught what you meant.  Will remove the old image tags.

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:1058]: If we're changing the usage, let's update the "usage" function by adding something like LVMS_IMAGE_MULTI_ARCH?
We would also need to update the rebase automation calling this script.

File: scripts/auto-rebase/rebase.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase.sh:190]: All this construct looks a bit complex, cat images, if with nested while, etc. Can we please simplify this code?

**Review by ggiguash — CHANGES_REQUESTED** [review]

**Comment by ggiguash** [comment]
@copejon, please update the rebase documentation in this PR to reflect the changes to the script.

File: scripts/auto-rebase/rebase.sh
Comment (dhellmann) [reply scripts/auto-rebase/rebase.sh:131]: Is `oc` available in the rebase script when it runs in CI?

File: scripts/auto-rebase/rebase.sh
Comment (dhellmann) [inline scripts/auto-rebase/rebase.sh:1055]: Does the automated job get this new argument?

File: assets/release/release-x86_64.json
Comment (ggiguash) [inline assets/release/release-x86_64.json:13]: Looks like we're missing the new images here

**Review by ggiguash — CHANGES_REQUESTED** [review]

File: assets/release/release-aarch64.json
Comment (pmtk) [inline assets/release/release-aarch64.json:3]: Could you use images from `last_rebase.sh`? To have this changes in sync with what we already have

File: scripts/auto-rebase/rebase.sh
Comment (pmtk) [inline scripts/auto-rebase/rebase.sh:1038]: So, job will have this argument hardcoded to `registry.redhat.io/lvms4/lvms-operator-bundle:v4.YY` depending on branch it runs for?  Like `release-4.12` -> `v4.12`, `main` -> `v4.13` (where it's released)?

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:131]: Yes, it's an existing prereq for the script.  This PR does not add new dependencies.

File: scripts/auto-rebase/rebase.sh
Comment (copejon) [reply scripts/auto-rebase/rebase.sh:1055]: Not yet.  To prevent breaking the job in CI, I intend to update CI after this is merged to update the env vars that are consumed by the wrapper script (rebase-script-automation.sh), and then glue it all together with an update to wrapper script to pass the env var thru. Hopefully, this will mean the rebase job doesn't see any downtime.

Comment (pmtk) [reply scripts/auto-rebase/rebase.sh:1055]: This line changes expectation that script will receive one more argument, so shouldn't we change job config so we have this available in container and then merge this change + changes in `rebase_job_entrypoint.sh`?
If you merge this first and not change `rebase_job_entrypoint.sh` then the script will fail expecting 4 args when given 3 or I missed something?

File: scripts/auto-rebase/rebase_job_entrypoint.sh
Comment (ggiguash) [inline scripts/auto-rebase/rebase_job_entrypoint.sh:26]: How is this going to work in main branch? I mean, we have no 4.13 release of LVMS yet, don't we?

Comment (pmtk) [inline scripts/auto-rebase/rebase_job_entrypoint.sh:26]: I think, in CI, release_amd64 is already a tag:
```
++ oc get configmap/release-release-images-latest -o yaml
++ yq '.data."release-images-latest.yaml"'
++ jq -r .metadata.name
+ release_amd64=4.13.0-0.nightly-2023-01-27-165107
```
so I guess you'd need to drop `.*:` from regexp.
Also, Giga is right - we might just hardcode 4.12 for now

File: assets/components/lvms/topolvm-controller_deployment.yaml
Comment (pmtk) [inline assets/components/lvms/topolvm-controller_deployment.yaml:12]: I don't see this file in:
```
update_lvms_manifests() {
    local src="${STAGING_DIR}/lvms/amd64"
    local dest="${REPOROOT}/assets/components/lvms"
    local manifests=(
          "topolvm-controller_rbac.authorization.k8s.io_v1_clusterrolebinding.yaml"
          "topolvm-controller_rbac.authorization.k8s.io_v1_clusterrole.yaml"
          "topolvm-controller_v1_serviceaccount.yaml"
          "topolvm-csi-provisioner_rbac.authorization.k8s.io_v1_clusterrole.yaml"
          "topolvm-csi-provisioner_rbac.authorization.k8s.io_v1_clusterrolebinding.yaml"
          "topolvm-csi-provisioner_rbac.authorization.k8s.io_v1_role.yaml"
          "topolvm-csi-provisioner_rbac.authorization.k8s.io_v1_rolebinding.yaml"
          "topolvm-csi-resizer_rbac.authorization.k8s.io_v1_clusterrole.yaml"
          "topolvm-csi-resizer_rbac.authorization.k8s.io_v1_clusterrolebinding.yaml"
          "topolvm-csi-resizer_rbac.authorization.k8s.io_v1_role.yaml"
          "topolvm-csi-resizer_rbac.authorization.k8s.io_v1_rolebinding.yaml"
          "topolvm-metrics_rbac.authorization.k8s.io_v1_role.yaml"
          "topolvm-metrics_rbac.authorization.k8s.io_v1_rolebinding.yaml"
          "topolvm-node-metrics_v1_service.yaml"
          "topolvm-node_rbac.authorization.k8s.io_v1_clusterrole.yaml"
          "topolvm-node_rbac.authorization.k8s.io_v1_clusterrolebinding.yaml"
          "topolvm-node_v1_serviceaccount.yaml"
          "topolvm.io_logicalvolumes.yaml"
    )
```
How that line changed? Shouldn't we have this file in the `manifests` array?

Comment (copejon) [reply assets/components/lvms/topolvm-controller_deployment.yaml:12]: unfortunately this file doesn't exist as part of the manifest bundle.  This deployment (and the daemonset) are both programmatically defined in the LVMS operator code.  We have a ticket open with LVMS team to add a render type behavior for these.

**Comment by pmtk** [comment]
This look good to me. We also tested changes using pj-rehearse with my fork and the rebase PR looks good as well.

/lgtm

File: assets/components/lvms/topolvm-node-metrics_v1_service.yaml
Comment (pmtk) [inline assets/components/lvms/topolvm-node-metrics_v1_service.yaml:17]: Sorry for digging this out, but it got me thinking...

We updated this file as it's not generated in runtime by the operator, but we're missing other part of the equation - labels in daemonset.
Looking at https://github.com/red-hat-storage/lvm-operator/blob/08b37d20e0711b72dabf884eeb77b228cde7c7fe/controllers/topolvm_node.go#L178-L182 it looks we also need to update labels in our topolvm-node_daemonset.yaml. 
Currently, we create Service, but it has nothing to point to?

Comment (pmtk) [reply assets/components/lvms/topolvm-node-metrics_v1_service.yaml:17]: While we wait for lvms team to provide us yamls, I think it could be faster to ask them to export ` getNodeDaemonSet()` and other functions so we could introduce yet another language to our rebase automation and generate these manifests ourselves :see_no_evil:

Comment (copejon) [reply assets/components/lvms/topolvm-node-metrics_v1_service.yaml:17]: As suggested, I generated the deployment and daemonset manifests (the only 2 not available in the operator bundle) and selectively merged them onto our current manifests.  I ignored upstream values for resource requests, since MicroShift runs with reduced values. Also ignored image refs since inject those those via go template.  Lastly, I left off the snapshotter container spec.

Comment (copejon) [reply assets/components/lvms/topolvm-node-metrics_v1_service.yaml:17]: > While we wait for lvms team to provide us yamls, I think it could be faster to ask them to export getNodeDaemonSet() and other functions so we could introduce yet another language to our rebase automation and generate these manifests ourselves 🙈

I think taking it one step further and adding a "render" command to the operator would be easier to manager from our perspective.

**Comment by pmtk** [comment]
Looking at updated manifests, I think we can drop labels with `lvms-operator` values, but we can do it in another PR

**Pull request merged**
